### ISSUE TITLE
compiling mlpack on cygwin

### ISSUE THREAD

**Issue by rpanwar2020** [mlpack/mlpack#1638]
I am trying to compile mlpack 3.0.4 on Cygwin64. I am getting the following errors when running cmake. Prior to invoking the cmake command below for installing mlpack, I installed armadillo as follows:
```
tar -xf armadillo-9.200.6.tar.xz
cd armadillo-9.200.6
mkdir build
cd build
cmake -DBUILD_SHARED_LIBS=OFF ..
make install
```
My cywin64 had BLAS and LAPACK already installed and so the armadillo install went ok. Please advise on how I can fix the mlpack cmake issue. Thank you!

[mlpack-cmake.txt](https://github.com/mlpack/mlpack/files/2735838/mlpack-cmake.txt)

**Comment by KimSangYeon-DGU** [comment]
@rpanwar2020 Did you read [this](https://github.com/mlpack/mlpack/wiki/WindowsBuild)??

**Comment by rcurtin** [comment]
It looks to me like mlpack is trying to use Armadillo symbols that depend on HDF5.  Since you are using the wrapper (libarmadillo.a), then basically either the wrapper needs to be linked against HDF5, or you need to comment out the `#define ARMA_USE_HDF5` in `armadillo_bits/config.hpp`.  I think if you make that change and rebuild Armadillo, everything should work fine.

**Comment by rpanwar2020** [comment]
@KimSangYeon-DGU and @rcurtin: thank you for the helpful pointers. I will first try to make mlpack work on cygwin since I use it for other projects. If I can't get cygwin to work, then I will look at MSVC option. 

Per the suggestion, I removed the HDF5 dependency. I did it by using -DDETECT_HDF5 in cmake invocation. However, it was also giving me unresolved symbols from blas and lapack. So I reinstalled blas and lapack on cygwin. After that, I resintalled armadillo. And then I tried recompiling mlpack as follows:

``` 
tar` xvzpf lapack-3.8.0
cd lapack-3.8.0
mkdir build
cd build
cmake -DBUILD_SHARED_LIBS=OFF ..
make install
```
```
tar -xf armadillo-9.200.6.tar.xz
cd armadillo-9.200.6
mkdir build
cd build
cmake -DBUILD_SHARED_LIBS=OFF -DDETECT_HDF5=false ..
make install
```
```
    tar xvzpf mlpack-3.0.4.tar.gz
    cd ml-3.0.4
    mkdir build
    cd build
    cmake -DBUILD_SHARED_LIBS=OFF -DUSE_OPENMP=OFF -DBUILD_PYTHON_BINDINGS=OFF ..
    make install
```


However, I am still getting unresolved symbols from blas and lapack during make install of mlpack:                                                                                  
```
[ 25%] Linking CXX executable ../../../../bin/mlpack_adaboost.exe
/usr/lib/libarmadillo.a(wrapper.cpp.o):wrapper.cpp:(.text+0x1): undefined reference to `sasum_'
/usr/lib/libarmadillo.a(wrapper.cpp.o):wrapper.cpp:(.text+0x1): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `sasum_'
/usr/lib/libarmadillo.a(wrapper.cpp.o):wrapper.cpp:(.text+0x11): undefined reference to `dasum_'
/usr/lib/libarmadillo.a(wrapper.cpp.o):wrapper.cpp:(.text+0x11): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `dasum_'

```

The new libblas.s and liblapack.a did get installed in /usr/local/lib and armadillo did find them when I ran cmake on armadillo. Partial output from that cmake:

```
-- *** Armadillo wrapper library will use the following libraries:
-- *** ARMA_LIBS = /usr/local/lib/libblas.a;/usr/local/lib/liblapack.a;/usr/lib/libarpack.dll.a

```

Somehow, when running cmake on mlpack, libarmadillo.a is not resolving some of the blas and lapack symbols. objdump of the new libblas.a shows sasum lives there:
```
objdump -t libblas.a  | grep sasum
sasum.f.o:     file format pe-x86-64
[  0](sec -2)(fl 0x00)(ty   0)(scl 103) (nx 1) 0x0000000000000000 sasum.f
[  2](sec  1)(fl 0x00)(ty  20)(scl   2) (nx 1) 0x0000000000000000 sasum_

```

**Comment by rcurtin** [comment]
Hmm, so I am not a Windows expert, but I do wonder if it might be worth avoiding the Armadillo wrapper by commenting out `#define ARMA_USE_WRAPPER` in `config.hpp`, and then mlpack's CMake configuration should link directly against the BLAS and LAPACK libraries.

**Comment by rcurtin** [comment]
:+1: Let me know if that works.  Even though I am not a Windows expert I can learn by proxy by knowing what worked for people. :)

**Comment by rpanwar2020** [comment]
Followed your suggestion about bypassing armadillo via the ARMA_USE_WRAPPER. That helped quench most of the blas and libpack errors. However, there were still some lingering errors with unresolved symbols in blas. So I am putting this mlpack on cygwin project on hold for the moment. I will get back to it later in the year and will update you.

**Comment by rcurtin** [comment]
Sounds good, glad that change helped you get closer.  If you come back to this project, happy to try and help more.  Maybe we might discover some improvements we can make to the CMake configuration that will make it easier for others. :+1:
